promq/autocomplete/earley/suggested_types_test.go

```diff
@@ -50,158 +50,389 @@ func TestEarleyItems(t *testing.T) {
+			inputString:  "foo and bar",
+			tokenPosList: []int{1, 2, 3},
+			expectedTypesList: [][]TokenType{
+				{ARITHMETIC, COMPARISION, SET, OFFSET_KW, LEFT_BRACKET, LEFT_BRACE, LEFT_PAREN, EOF},
+				{NUM, METRIC_ID, AGGR_OP, FUNCTION_SCALAR_ID, FUNCTION_VECTOR_ID, GROUP_KW, LEFT_PAREN},
+				{OFFSET_KW, LEFT_PAREN, LEFT_BRACE, LEFT_BRACKET, SET, COMPARISION, ARITHMETIC, EOF},
+			},
+		},
+		{
+			name:         "Binary Expression - one_to_one vector match with arithmetic operator",
+			inputString:  "foo * on(test,) bar",
+			tokenPosList: []int{2, 3, 4, 5, 6, 7, 8},
+			expectedTypesList: [][]TokenType{
+				{NUM, METRIC_ID, AGGR_OP, FUNCTION_SCALAR_ID, FUNCTION_VECTOR_ID, GROUP_KW, LEFT_PAREN},
+				{LEFT_PAREN},
+				{RIGHT_PAREN, METRIC_LABEL_SUBTYPE},
+				{COMMA, RIGHT_PAREN},
+				{RIGHT_PAREN, METRIC_LABEL_SUBTYPE},
+				{GROUP_SIDE, NUM, METRIC_ID, FUNCTION_VECTOR_ID, FUNCTION_SCALAR_ID, AGGR_OP, LEFT_PAREN},
+				{SET, OFFSET_KW, LEFT_PAREN, LEFT_BRACKET, LEFT_BRACE, COMPARISION, ARITHMETIC, EOF},
+			},
+		},
+		{
+			name:         "Binary Expression - one_to_one vector match with set operator",
+			inputString:  "foo and on(test,) bar",
+			tokenPosList: []int{7, 8},
+			expectedTypesList: [][]TokenType{
+				{NUM, METRIC_ID, FUNCTION_VECTOR_ID, FUNCTION_SCALAR_ID, AGGR_OP, LEFT_PAREN},
+				{SET, OFFSET_KW, LEFT_PAREN, LEFT_BRACKET, LEFT_BRACE, COMPARISION, ARITHMETIC, EOF},
+			},
+		},
+		{
+			name:         "Binary Expression - one_to_many vector match",
+			inputString:  "foo / on(test,blub) group_left (bar,) bar",
+			tokenPosList: []int{8, 9, 10, 11, 12, 13, 14},
+			expectedTypesList: [][]TokenType{
+				{GROUP_SIDE, NUM, METRIC_ID, FUNCTION_VECTOR_ID, FUNCTION_SCALAR_ID, AGGR_OP, LEFT_PAREN},
+				{LEFT_PAREN, NUM, METRIC_ID, FUNCTION_VECTOR_ID, FUNCTION_SCALAR_ID, AGGR_OP},
+				{METRIC_LABEL_SUBTYPE, RIGHT_PAREN, NUM, METRIC_ID, LEFT_PAREN, FUNCTION_VECTOR_ID, FUNCTION_SCALAR_ID, AGGR_OP, UNARY_OP},
+				{COMMA, RIGHT_PAREN, OFFSET_KW, LEFT_PAREN, COMPARISION, ARITHMETIC, LEFT_BRACE, SET},
+				{METRIC_LABEL_SUBTYPE, RIGHT_PAREN},
+				{NUM, METRIC_ID, FUNCTION_VECTOR_ID, FUNCTION_SCALAR_ID, AGGR_OP, LEFT_PAREN},
+				{SET, OFFSET_KW, LEFT_PAREN, LEFT_BRACKET, LEFT_BRACE, COMPARISION, ARITHMETIC, EOF},
+			},
+		},
+		{
+			name:         "Metric Expression - with labels",
+			inputString:  "metric_name{label1='foo', label2='bar'}",
+			tokenPosList: []int{1, 2, 3, 4, 5, 6, 10},
+			expectedTypesList: [][]TokenType{
+				{EOF, LEFT_BRACE, OFFSET_KW, LEFT_BRACKET, LEFT_PAREN, COMPARISION, ARITHMETIC, SET},
+				{METRIC_LABEL_SUBTYPE},
+				{LABELMATCH},
+				{STRING},
+				{RIGHT_BRACE, COMMA},
+				{METRIC_LABEL_SUBTYPE},
+				{EOF, OFFSET_KW, LEFT_BRACKET, COMPARISION, ARITHMETIC, SET},
+			},
+		},
+		{
+			name:         "Metric Expression - metric name contains `:`",
+			inputString:  "metric:name{label1='foo', label2='bar'}",
+			tokenPosList: []int{1, 2, 3, 4, 5, 6, 10},
```

Comment:
This is a highly personal preference, but I feel these would read a little better if rather than a separate list we used a map:

```
    expectedTypesFromParsePosMap: map[int][]TokenType
```

Reply:
I like the map. Changed. Thanks.